File: I_Treasure_Hunt.cpp
#include<bits/stdc++.h>
#define ll long long
using namespace std;
ll dp[1005][1005];
ll knapsack(ll n,ll s,ll w[],ll v[]){
    if(s==0 || n==0) return 0;
    if(dp[n][s] != -1){
        return dp[n][s];
    }
    if(w[n-1]<=s){
        ll op1=knapsack(n-1,s-w[n-1],w,v)+v[n-1];
        ll op2=knapsack(n-1,s,w,v);
        return dp[n][s]=max(op1,op2);
    }
    else{
        return dp[n][s]=knapsack(n-1,s,w,v);
    }
}
int main()
{   
    ll n,s; cin>>n>>s;
    ll w[n],v[n];
    for(int i=0;i<n;i++){
        cin>>w[i]>>v[i];
    }
    for(int i=0;i<=n;i++){
        for(int j=0;j<=s;j++){
            dp[i][j]=-1;
        }
    }
    cout<<knapsack(n,s,w,v);
    return 0;
}

File: I_Treasure_Hunt.py
# Python implementation of the knapsack problem

def knapsack(n, s, w, v, dp):
    # Base case: no items or no remaining weight capacity
    if s == 0 or n == 0:
        return 0
    
    # If this subproblem is already solved, return the cached result
    if dp[n][s] != -1:
        return dp[n][s]
    
    # If the item can fit in the knapsack
    if w[n - 1] <= s:
        # Option 1: Include the item
        op1 = knapsack(n - 1, s - w[n - 1], w, v, dp) + v[n - 1]
        # Option 2: Exclude the item
        op2 = knapsack(n - 1, s, w, v, dp)
        # Store the maximum value of the two options
        dp[n][s] = max(op1, op2)
    else:
        # If item cannot be included, exclude it
        dp[n][s] = knapsack(n - 1, s, w, v, dp)
    
    return dp[n][s]

# Main function to initialize variables and call knapsack
if __name__ == "__main__":
    # Input number of items and knapsack capacity
    n = int(input())
    s = int(input())
    
    # Input weights and values
    w = []
    v = []
    for i in range(n):
        weight, value = map(int, input().split())
        w.append(weight)
        v.append(value)
    
    # Initialize dp array with -1 for memoization
    dp = [[-1 for _ in range(s + 1)] for _ in range(n + 1)]
    result = knapsack(n, s, w, v, dp)
    print(result)
